pkg/daemon/server/service/pipeline_watermark_query.go

```diff
@@ -0,0 +1,107 @@
+// Package service is built for querying metadata and to expose it over daemon service.
+package service
+
+import (
+	"context"
+	"fmt"
+	"github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	"github.com/numaproj/numaflow/pkg/apis/proto/daemon"
+	"github.com/numaproj/numaflow/pkg/isbsvc"
+	"github.com/numaproj/numaflow/pkg/isbsvc/clients"
+	"github.com/numaproj/numaflow/pkg/shared/logging"
+	"github.com/numaproj/numaflow/pkg/watermark/fetch"
+	"github.com/numaproj/numaflow/pkg/watermark/generic"
+	"github.com/numaproj/numaflow/pkg/watermark/store/jetstream"
+	"go.uber.org/zap"
+	"time"
+)
+
+// watermarkFetchers used to store watermark metadata for propagation
+type watermarkFetchers struct {
+	fetchMap           map[string]fetch.Fetcher
+	isWatermarkEnabled bool
+}
+
+// newVertexWatermarkFetcher creates a new instance of watermarkFetchers. This is used to populate a map of vertices to
+// corresponding fetchers. These fetchers are tied to the incoming edge buffer of the current vertex (Vn), and read the
+// watermark propagated by the vertex (Vn-1). As each vertex has one incoming edge, for the input vertex we read the source
+// data buffer.
+func newVertexWatermarkFetcher(pipeline *v1alpha1.Pipeline) *watermarkFetchers {
+	ctx := context.Background()
+	log := logging.FromContext(ctx)
```

Comment:
This will be a new created log which has not context.

A better approach in this function would be returning `(*xxx, error )` instead of printing logs.